includes/Core/Util/User_Input_Settings.php

```diff
@@ -126,19 +126,27 @@ function( $setting ) {
 	 * @return array|WP_Error User input settings.
 	 */
 	private function sync_with_proxy( $settings = null ) {
-		$user_input_settings_url  = $this->authentication->get_google_proxy()->url( Google_Proxy::USER_INPUT_SETTINGS_URI );
+		$user_input_settings_url = $this->authentication->get_google_proxy()->url( Google_Proxy::USER_INPUT_SETTINGS_URI );
+		$creds                   = $this->authentication->credentials()->get();
+
 		$user_input_settings_args = array(
 			'headers' => array(
 				'Authorization' => 'Bearer ' . $this->authentication->get_oauth_client()->get_access_token(),
+				'Content-Type'  => 'application/json',
+			),
+			'body'    => array(
+				'site_id'     => $creds['oauth2_client_id'],
+				'site_secret' => $creds['oauth2_client_secret'],
 			),
 		);
 
 		if ( ! empty( $settings ) ) {
-			$user_input_settings_args['headers']['Content-Type'] = 'application/json';
-			$user_input_settings_args['body']                    = wp_json_encode( $settings );
+			$user_input_settings_args['body']['settings']       = $settings;
+			$user_input_settings_args['body']['client_user_id'] = get_current_user_id();
 		}
 
-		$response = wp_remote_post( $user_input_settings_url, $user_input_settings_args );
+		$response = wp_remote_post( $user_input_settings_url, wp_json_encode( $user_input_settings_args ) );
+
 		if ( is_wp_error( $response ) ) {
 			return $response;
 		}
```

Comment:
I cannot mark the code I actually intend to here since you actually didn't touch it, but it's worth improving because right now the error messaging here is not great because it doesn't provide any details that may be present in the API response.
1. Can you move the response body decoding (lines 163-164) above the check for the response code (lines 154-155)?
2. Then, in that clause, add another nested `if ( is_array( $settings ) && ! empty( $settings['error'] ) )`. If that is set (it's an error code), return a new `WP_Error` where you use that value as error code and something like `sprintf( __( 'User input settings request failed with error: %s', 'google-site-kit' ), $settings['error'] )` as message.

This way you'll be able to tell from the client side _what_ actually went wrong in most cases.

Let's keep the currently used unprecise `WP_Error` only for when we cannot get more details from the API response like I described above.

Reply:
Hey Felix, I've modified this based on your suggestions to [surface errors from the API more clearly](https://github.com/google/site-kit-wp/pull/2489/files#diff-f2d89ed2b380a2e93a3a19fd41ec17471f836fcf9b2ffc8c0bc44ae06ab672ddR162-R169).

Reply:
This looks solid, the only thing worth improving would be to use the `error` code as error code for the `WP_Error` as well, as that provides more insight than a generic `user_input_settings_request`.